common/subtitle-coloring/subtitle-coloring.ts

```diff
@@ -0,0 +1,1145 @@
+import { Fetcher, ColoredSubtitleModel } from '@project/common';
+import { Anki } from '@project/common/anki';
+import {
+    AsbplayerSettings,
+    DictionaryAnkiTreatSuspended,
+    DictionarySubtitleAppearance,
+    DictionaryTrack,
+    getFullyKnownTokenStatus,
+    TokenMatchStrategy,
+    TokenMatchStrategyPriority,
+    TokenStatus,
+    TokenStyling,
+} from '@project/common/settings';
+import { SubtitleCollection } from '@project/common/subtitle-collection';
+import { arrayEquals, filterAsync, inBatches } from '@project/common/util';
+import { Yomitan } from '@project/common/yomitan/yomitan';
+
+const TOKEN_CACHE_BUILD_AHEAD = 10;
+const TOKEN_CACHE_BATCH_SIZE = 1; // Processing more than 1 at a time is slower
+const TOKEN_CACHE_ERROR_REFRESH_INTERVAL = 10000;
+const ANKI_RECENTLY_MODIFIED_INTERVAL = 10000;
+const HAS_LETTER_REGEX = /\p{L}/u;
+
+export class SubtitleColoring {
```

Comment:
Suggestion: move subtitles and the color state closer together, instead of maintaining separate colored and uncolored copies of `subtitles`. You could make `SubtitleCollection` into an interface with `bind` and `unbind` methods, and have a colored subtitle collection that wraps the original one.

I would expect that this simplifies any code below that requires knowledge of the currently-showing subtitles - because you already know as soon as the slice is requested. It should also simplify any code that is instantiating the `SubtitleColoring` separately from the `SubtitleCollection`.